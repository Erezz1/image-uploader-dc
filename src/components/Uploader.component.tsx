import { useState, useCallback, FC } from 'react';
import { useDropzone } from 'react-dropzone'
import Image from 'next/image';

import { validFile } from '@/helpers';
import { uploadFile } from '@/services';

import styles from './uploader-component.module.scss';
import image from '../resources/image.80539fd7.svg';
import imageUp from '../resources/image-upload.svg';

const UploaderComponent = () => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [url, setUrl] = useState<string | null>(null);

  const handleUploadFile = async (file: File) => {
    setIsLoading(true);
    const url = await uploadFile(file);
    setUrl(url);
    setIsLoading(false);
  }

  return (
    <div className={styles.container}>
      {
        isLoading ?
          <Loading /> :
          !url ?
            <Upload uploadFile={handleUploadFile} /> :
            <Uploaded imageUrl={url} />
      }
    </div>
  )
}

interface UploadProps {
  uploadFile: (file: File) => void;
}
const Upload: FC<UploadProps> = ({ uploadFile }) => {
  const onDrop = useCallback((acceptedFiles: File[]) => {
    if (!validFile(acceptedFiles[0])) return;
    uploadFile(acceptedFiles[0]);
  }, []);
  const { getRootProps, getInputProps, isDragActive, open } = useDropzone({ onDrop });

  return (
    <>
      <h2>Upload your image</h2>
      <p>File should be Jpeg, Png,...</p>

      <div className={styles.dropzone} {...getRootProps()}>
        <input {...getInputProps()} />
        {isDragActive ? (<>
          <Image
            width="115"
            height="90"
            alt="drag-here-image"
            src={imageUp}
          />
          <p>Upload...</p>
        </>) : (<>
          <Image
            src={image}
            alt="upload-image"
          />
          <p>Drag & Drop your image here</p>
        </>)}
      </div>

      <p>Or</p>
      <button onClick={open}>Choose a file</button>
    </>
  )
}

const Loading = () => (
  <>
    <h3 style={{ alignSelf: 'flex-start' }}>
      Uploading...
    </h3>
    <span className={styles.loader}></span>
  </>
)

interface UploadedProps {
  imageUrl: string;
}
const Uploaded: FC<UploadedProps> = ({ imageUrl }) => {
  const handleCopyUrl = () => navigator.clipboard.writeText(imageUrl);

  return (
    <>
      <span className={`material-symbols-outlined ${styles.icon}`}>
        check_circle
      </span>
      <h2>Uploaded Successfully!</h2>
      <img
        className={styles.image}
        src={imageUrl}
        alt="image-uploaded"
      />
      <div className={styles.text}>
        {imageUrl}
        <button
          className={styles.copy}
          onClick={handleCopyUrl}
        >
          Copy Link
        </button>
      </div>
    </>
  )
}

export default UploaderComponent;
