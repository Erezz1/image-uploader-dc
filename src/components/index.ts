export { default as UploaderComponent } from './Uploader.component';
export * from './Uploader.component';
