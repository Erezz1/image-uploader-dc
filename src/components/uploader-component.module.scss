.container {
  width: 90%;
  max-width: 400px;
  padding: 2.5rem 1.5rem;
  background-color: #FFFFFF;
  box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.1);
  border-radius: 12px;
  display: flex;
  align-items: center;
  flex-direction: column;
  color: #4F4F4F;

  h2,
  h3 {
    margin: 0;
    font-size: 18px;
    font-weight: normal;
    letter-spacing: -0.035em;
  }

  h3 {
    font-size: 16px;
    margin-bottom: 1.5rem;
  }

  p {
    font-size: 12px;
    font-weight: normal;
    letter-spacing: -0.035em;
    color: #828282;
  }

  button {
    padding: 8px 1rem;
    background-color: #2F80ED;
    border-radius: 8px;
    color: #FFFFFF;
    border: none;
    cursor: pointer;
    transition: all 300ms;
    margin-top: 0.5rem;

    &:hover {
      background-color: #2564b5;
    }

    &:active {
      background-color: #4581ce;
    }
  }

  .dropzone {
    width: 100%;
    background-color: #F6F8FB;
    border: 1px dashed #97BEF4;
    border-radius: 12px;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 2rem;
    margin: 1rem 0;
    gap: 1rem;

    p {
      color: #BDBDBD
    }
  }

  .image {
    width: 100%;
    height: 200px;
    object-fit: cover;
    margin: 1.5rem 0;
    border-radius: 12px;
  }

  .text {
    position: relative;
    width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    padding: 11px 7px;
    font-size: 12px;
    background: #F6F8FB;
    border: 1px solid #E0E0E0;
    border-radius: 8px;
  }

  .copy {
    position: absolute;
    right: 4px;
    top: -4px;
  }
}

.loader {
  display: block;
  position: relative;
  height: 6px;
  width: 100%;
  border-radius: 10px;
  background-color: #F2F2F2;
  overflow: hidden;
}

.loader::after {
  content: '';
  width: 35%;
  height: 100%;
  background: #2F80ED;
  position: absolute;
  top: 0;
  left: 0;
  box-sizing: border-box;
  animation: animloader 1.75s linear infinite;
}

.icon {
  font-variation-settings:
    'FILL' 1,
    'wght' 400,
    'GRAD' 0,
    'opsz' 48;
  color: #219653;
  font-size: 3rem;
  margin-bottom: 11px;
}

@keyframes animloader {
  0% {
    left: 0;
    transform: translateX(-100%);
  }

  100% {
    left: 100%;
    transform: translateX(0%);
  }
}