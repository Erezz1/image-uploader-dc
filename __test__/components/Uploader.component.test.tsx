import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Loading, Uploaded, Upload, UploaderComponent } from '@/components';

afterEach(cleanup);

describe('Test in Uploader.component.tsx', () => {
  it('<UploaderComponent/> To do match in the snapshot', async () => {
    const { container } = render(<UploaderComponent />);
    expect(container).toMatchSnapshot();
  });

  it('should to show Dropzone component first', () => {
    render(<UploaderComponent />);
    const dropzone = screen.getAllByRole('presentation');
    expect(dropzone).toBeTruthy();
  });
});

describe('Test in component <Loading />', () => {
  test('should to do match in the snapshot', () => {
    const { container } = render(<Loading />);
    expect(container).toMatchSnapshot();
  });

  it('should show the message: "Uploading..."', () => {
    render(<Loading />);
    const text = screen.getByRole('heading', { level: 3 });
    expect(text.textContent).toBe('Uploading...');
  });
});

describe('Test in component <Uploaded />', () => {
  const url = 'https://images.unsplash.com/photo-1678737169727-a3236885e763?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=846&q=80';

  test('should to do match in the snapshot', () => {
    const { container } = render(<Uploaded imageUrl={url} />);
    expect(container).toMatchSnapshot();
  });

  it('should to do match the image url, p tag and test url', () => {
    render(<Uploaded imageUrl={url} />);
    const p = screen.getByText(url);
    const img = screen.getByRole('img');

    expect(p.textContent).toBe(url);
    expect(img.src).toBe(url);
    expect(p.textContent).toBe(img.src);
  });
});

describe('Test in component <Upload />', () => {
  const uploadFile = jest.fn();
  const file = new File(['test'], 'test.png', { type: 'image/png' });

  test('should to do match in the snapshot', () => {
    const { container } = render(<Upload uploadFile={uploadFile} />);
    expect(container).toMatchSnapshot();
  });

  it('should the input tag have a file', () => {
    render(<Upload uploadFile={uploadFile} />);
    const input = screen.getByTestId('drop-input');

    fireEvent.input(input, { target: { file: [file] } });

    expect(input.file).toHaveLength(1);
    expect(input.file[0]).toEqual(file);
  });
});
