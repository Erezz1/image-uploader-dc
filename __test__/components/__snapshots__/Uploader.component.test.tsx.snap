// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in Uploader.component.tsx <UploaderComponent/> To do match in the snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <h2>
      Upload your image
    </h2>
    <p>
      File should be Jpeg, Png,...
    </p>
    <div
      class="dropzone"
      role="presentation"
      tabindex="0"
    >
      <input
        data-testid="drop-input"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
      <img
        alt="upload-image"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <p>
        Drag & Drop your image here
      </p>
    </div>
    <p>
      Or
    </p>
    <button>
      Choose a file
    </button>
  </div>
</div>
`;

exports[`Test in component <Loading /> should to do match in the snapshot 1`] = `
<div>
  <h3
    style="align-self: flex-start;"
  >
    Uploading...
  </h3>
  <span
    class="loader"
  />
</div>
`;

exports[`Test in component <Upload /> should to do match in the snapshot 1`] = `
<div>
  <h2>
    Upload your image
  </h2>
  <p>
    File should be Jpeg, Png,...
  </p>
  <div
    class="dropzone"
    role="presentation"
    tabindex="0"
  >
    <input
      data-testid="drop-input"
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <img
      alt="upload-image"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
    <p>
      Drag & Drop your image here
    </p>
  </div>
  <p>
    Or
  </p>
  <button>
    Choose a file
  </button>
</div>
`;

exports[`Test in component <Uploaded /> should to do match in the snapshot 1`] = `
<div>
  <span
    class="material-symbols-outlined icon"
  >
    check_circle
  </span>
  <h2>
    Uploaded Successfully!
  </h2>
  <img
    alt="image-uploaded"
    class="image"
    src="https://images.unsplash.com/photo-1678737169727-a3236885e763?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=846&q=80"
  />
  <div
    class="text"
  >
    <p>
      https://images.unsplash.com/photo-1678737169727-a3236885e763?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=846&q=80
    </p>
    <button
      class="copy"
    >
      Copy Link
    </button>
  </div>
</div>
`;
